src/security/PeerOptions.cc

```diff
@@ -62,6 +62,13 @@ Security::PeerOptions::parse(const char *token)
         parseOptions();
     } else if (strncmp(token, "cipher=", 7) == 0) {
         sslCipher = SBuf(token + 7);
+#if USE_OPENSSL
+        // everything okay
+#elif USE_GNUTLS
+        debugs(3, DBG_PARSE_NOTE(1), "WARNING: SSL cipher= not supported. Uses GnuTLS tls-options= parameters instead.");
```

Comment:
Why not throw a configuration error?

Reply:
Because this is part of a complex line with many options that need different validation. The exception handling in Squid does not yet cope with this type of config line and the "-k parse" need to display more than one error on a single config line.

Reply:
> The exception handling in Squid does not yet cope with this type of config line

AFAICT, existing exception handling will handle the proposed exception just fine. That handling can be improved further, of course, but it is parsing and configuration code that blocks progress in most cases, not exception handling code.

> the "-k parse" need to display more than one error on a single config line.

It does not have to. There is no rule that `-k parse` has to detect all existing errors or even multiple existing errors. Such rule would be impractical to support in general because of the inter-dependencies between configuration pieces. When configuration errors are present, `-k parse` should detect at least one such error. And whether the errors are located on one line or multiple lines is completely irrelevant to the `-k parse` interface. Moreover, to minimize false errors and runtime-affecting bugs, we should be conservative -- when there is a risk of an error sipping through other checks, we should throw when the error is discovered. This is one of those cases AFAICT.

Reply:
* this cipher= option is documented by the WARNING and by the release notes as not having GnuTLS support (via lack of tls- prefix).

 * squid.conf clearly documents even OpenSSL itself as silently ignoring settings configured with this option if the ciphers are not usable. eg, for (EC)DH without a dhparams value.

Neither of which is great, but fixing UI behaviour is out of scope. IMO for long-term fix it is more efficient to drop this option entirely and use options= to set the ciphers in a library-agnostic way - which is also not suitable for this v5 targeted PR.

Reply:
> fixing UI behaviour is out of scope

It is highly debatable whether we are talking about a "fix" or "correct implementation" here. However, even if we were talking about a fix, I see no valid reason to place UI improvements outside this PR scope, especially when the PR is adding or creating those UI problems and the fix appears to be simple.

> not suitable for this v5 targeted PR

IMO, "targeting vX" is not a valid excuse for a PR submitted against the master branch. This PR is submitted against the master branch. It should be evaluated as such. If some of the changes are not appropriate for v5 (and that is a big "if"!), then they can be removed or adjusted when backporting to v5.

If really necessary, we could make an exception by splitting this PR into two parts: One part implements v5-suitable changes and the other does the right thing for master/v6. No pressing need for such an exception has been demonstrated AFAICT.

P.S. If a PR has been allowed to go into v5 around v5 branching time, it can modify v5 UI. Exceptions for valuable PRs that did not have enough time to merge before v5 was branched do not stop at the UI boundary. In this particular case, there is more harm in creating that boundary than in letting correct handling of misconfigurations through into v5.